vllm/entrypoints/openai/protocol.py

```diff
@@ -33,6 +41,64 @@
 assert _LONG_INFO.max == _MOCK_LONG_INFO.max
 
 
+class CustomChatCompletionMessageParam(TypedDict, total=False):
+    """Enables custom roles in the Chat Completion API."""
+    role: Required[str]
+    """The role of the message's author."""
+
+    content: Union[str, List[ChatCompletionContentPartParam]]
+    """The contents of the message."""
+
+    name: str
+    """An optional name for the participant.
+
+    Provides the model information to differentiate between participants of the
+    same role.
+    """
+
+    tool_call_id: Optional[str]
+
+    tool_calls: Optional[List[dict]]
+
+
+@final  # So that it should be compatible with Dict[str, str]
+class ConversationMessage(TypedDict, total=False):
+    role: str
+    content: Optional[str]
+    tool_call_id: Optional[str]
+    name: Optional[str]
+    tool_calls: Optional[List]
+
+
+class CustomChatCompletionContentPartParam(TypedDict, total=False):
+    __pydantic_config__ = ConfigDict(extra="allow")  # type: ignore
+
+    type: Required[str]
+    """The type of the content part."""
+
+
+class AudioURL(TypedDict, total=False):
+    url: Required[str]
+    """
+    Either a URL of the audio or a data URL with base64 encoded audio data.
+    """
+
+
+class ChatCompletionContentPartAudioParam(TypedDict, total=False):
+    audio_url: Required[AudioURL]
+
+    type: Required[Literal["audio_url"]]
+    """The type of the content part."""
+
+
+ChatCompletionContentPartParam = Union[OpenAIChatCompletionContentPartParam,
+                                       ChatCompletionContentPartAudioParam,
+                                       CustomChatCompletionContentPartParam]
+
+ChatCompletionMessageParam = Union[OpenAIChatCompletionMessageParam,
+                                   CustomChatCompletionMessageParam]
+
+
```

Comment:
These classes were previously moved from here to `chat_utils.py` to support offline chat API. Is there a reason why they have been moved back here?

Reply:
Yes - `protocol.py` is where type definitions relevant to compatibility with the OpenAI protocol go. These had previously been in `protocol.py`. Someone refactored it to be in `chat_utils.py` while I was working on this PR, but logically they belong in `protocol.py`.

Reply:
It's outdated again now.